storm-core/src/jvm/org/apache/storm/daemon/Task.java

```diff
@@ -0,0 +1,246 @@
+        Grouper grouping = componentGrouping.get(targetComponent);
+        if (null == grouping) {
+            outTaskId = null;
+        }
+        if (grouping != null && grouping != GrouperFactory.DIRECT) {
+            throw new IllegalArgumentException("Cannot emitDirect to a task expecting a regular grouping");
+        }
+        new EmitInfo(values, stream, taskId, Collections.singletonList(outTaskId)).applyOn(userTopologyContext);
+        try {
+            if (emitSampler.call()) {
+                executorStats.emittedTuple(stream);
+                if (null != outTaskId) {
+                    executorStats.transferredTuples(stream, 1);
+                }
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+        if (null != outTaskId) {
+            return Collections.singletonList(outTaskId);
+        }
+        return null;
+    }
+
+    public List<Integer> getOutgoingTasks(String stream, List<Object> values) {
+        if (debug) {
+            LOG.info("Emitting: {} {} {}", componentId, stream, values);
+        }
+        List<Integer> outTasks = new ArrayList<>();
+        if (!streamComponentToGrouper.containsKey(stream)) {
+            throw new IllegalArgumentException("Unknown stream ID: " + stream);
+        }
+        if (null != streamComponentToGrouper.get(stream)) {
+            // null value for __system
+            for (Grouper grouper : streamComponentToGrouper.get(stream).values()) {
+                if (grouper == GrouperFactory.DIRECT) {
+                    throw new IllegalArgumentException("Cannot do regular emit to direct stream");
+                }
+                List<Integer> compTasks = grouper.outTasks(taskId, values, loadMapping);
+                outTasks.addAll(compTasks);
+            }
+        }
+        new EmitInfo(values, stream, taskId, outTasks).applyOn(userTopologyContext);
+        try {
+            if (emitSampler.call()) {
+                executorStats.emittedTuple(stream);
+                executorStats.transferredTuples(stream, outTasks.size());
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+        return outTasks;
+    }
+
+    public Tuple getTuple(String stream, List values) {
+        return new TupleImpl(systemTopologyContext, values, systemTopologyContext.getThisTaskId(), stream);
+    }
+
+    public Integer getTaskId() {
+        return taskId;
+    }
+
+    public String getComponentId() {
+        return componentId;
+    }
+
+    public TopologyContext getUserContext() throws IOException {
+        return userTopologyContext;
+    }
+
+    public Object getTaskObject() {
+        return taskObject;
+    }
+
+    public BuiltinMetrics getBuiltInMetrics() {
+        return builtInMetrics;
+    }
+
+    private TopologyContext mkTopologyContextBuilder(StormTopology topology) throws IOException {
```

Comment:
It looks like `mk-topology-context-builder`, `system-topology-context`, and `user-topology-context` all got merged into this.  It probably would be best to just call it `mkTopologyContext`, and eventually have a Constructor on `TopologyContext` that just takes a StormTopology.